server/app/auth/oidc/OidcProvider.java

```diff
@@ -0,0 +1,185 @@
+
+  /*
+   * The OIDC Provider Name (optional).
+   */
+  protected abstract Optional<String> getProviderName();
+
+  /*
+   * Provide the profile adaptor that should be used.
+   */
+  public abstract ProfileCreator getProfileAdapter(OidcConfiguration config, OidcClient client);
+
+  /*
+   * The OIDC Client ID.
+   */
+  protected abstract String getClientID();
+
+  /*
+   * The OIDC Client Secret.
+   */
+  protected abstract String getClientSecret();
+
+  /*
+   * The OIDC Discovery URI.
+   */
+  protected abstract String getDiscoveryURI();
+
+  /*
+   * The OIDC Response Mode.
+   */
+  protected abstract String getResponseMode();
+
+  /*
+   * The OIDC Response Type.
+   */
+  protected abstract String getResponseType();
+
+  /*
+   * Default scopes for this OIDC implementation.
+   */
+  protected abstract ImmutableList<String> getDefaultScopes();
+
+  /*
+   * Any application.conf-defined extra scopes for this OIDC implementation.
+   */
+  protected abstract ImmutableList<String> getExtraScopes();
+
+  /*
+   * Helper function for retriving values from the application.conf,
+   * prepended with "<attributePrefix>."
+   */
+  protected final Optional<String> getConfigurationValue(String suffix) {
+    String name = attributePrefix() + "." + suffix;
+    return getBaseConfigurationValue(name);
+  }
+
+  /*
+   * Helper function for retriving values from the application.conf,
+   * prepended with "<attributePrefix>."
+   */
+  protected final Optional<String> getBaseConfigurationValue(String name) {
+    if (configuration.hasPath(name)) {
+      return Optional.ofNullable(configuration.getString(name));
+    }
+    return Optional.empty();
+  }
+
+  protected String getCallbackURL() {
+    return baseUrl + "/callback";
+  }
+
+  private final String getScope() {
+    // Scopes are the other things that we want from the OIDC endpoint
+    // (needs to also be configured on provider side).
+    ImmutableList<String> allClaims =
+        ImmutableList.<String>builder().addAll(getDefaultScopes()).addAll(getExtraScopes()).build();
+    return String.join(" ", allClaims);
+  }
+
+  @Override
+  public OidcClient get() {
+    String clientID = getClientID();
+    String clientSecret = getClientSecret();
+    String discoveryURI = getDiscoveryURI();
+    String responseMode = getResponseMode();
+    String responseType = getResponseType();
+    String callbackURL = getCallbackURL();
+    String scope = getScope();
+    var missing =
+        ImmutableList.of(
+                clientID, clientSecret, discoveryURI, responseMode, responseType, callbackURL)
+            .stream()
+            .map(Strings::nullToEmpty);
+    // Check that none are null or blank.
+    if (missing.anyMatch(String::isBlank)) {
+      logger.error(
+          String.format(
+              "Can't get OIDC client - Missing some required Provider data: "
+                  + "clientID=%s, "
+                  + "clientSecret=%s, "
```

Comment:
Blocking: I'd avoid logging the client secret here so we don't have to think about whether the person with logs access has access to the secret or not. In this case, I'd print "<redacted>" if it's non-empty.